Add basic syncing to firebase where the local device is the source of truth.

miguelandres opened the pull request:
-Each device gets its own object in firebase so devices are isolated and there is no cross-device syncing.
-The root object for each device is “remixer/<DEVICE_REMOTE_ID>”
-DEVICE_REMOTE_ID is the first 8 digits of a randomly generated UUID (not personally identifiable).
-DEVICE_REMOTE_ID is persisted across application restarts until data is cleared. That will allow to reuse the same URL across different runs targeting the same app/device combination.
-Deserialization is done manually, because type erasure in Java makes it near impossible to deal with types. This way the remixer type system remains extensible.

Commit:
Add dummy configuration for Firebase Services.

This allows us to compile adding firebase dependencies, but doesn’t expose any actual sensitive data.

Commit:
Add google-services.json to gitignore.

Commit:
Add basic syncing to firebase where the local device is the source of truth.

-Each device gets its own object in firebase so devices are isolated and there is no cross-device syncing.
-The root object for each device is “remixer/<DEVICE_REMOTE_ID>”
-DEVICE_REMOTE_ID is the first 8 digits of a randomly generated UUID (not personally identifiable).
-DEVICE_REMOTE_ID is persisted across application restarts until data is cleared. That will allow to reuse the same URL across different runs targeting the same app/device combination.
-Deserialization is done manually, because type erasure in Java makes it near impossible to deal with types. This way the remixer type system remains extensible.

Commit:
Add play services to travis config

Commit:
Emulate Firebase quickstart travis config file

pingpongboss commented on .travis.yml:
```diff
@@ -15,13 +15,27 @@ android:
+cache:
+  directories:
+    - $HOME/.gradle/caches/
+    - $HOME/.gradle/wrapper/
```
Can you explain all these changes?

miguelandres replied:
I'm following the example in this canonical quickstart. https://github.com/firebase/quickstart-android/blob/master/.travis.yml

I definitely had to add the licenses and other dependencies, travis was not finding firebase libraries at all without those.

Caching, in particular, is also something I've been meaning to investigate, it seems to be working fine for them (and us in these few runs), so it may be a bit of a improvement regarding how long it takes to run CI.

pingpongboss commented on remixer_example/google-services.json:
```diff
@@ -0,0 +1,42 @@
+{
```
i thought this file is gitignored?

miguelandres replied:
I had to add a dummy version, otherwise the build fails if you apply plugin com.google.gms.google-services.

I need it to build even if it doesn't connect to firebase.

miguelandres replied:
I was thinking gitignoring it after adding it would do the trick so git would ignore it, but I think it does nothing *sigh* So I have to manually remember to never add this file.

pingpongboss replied:
That's fine. Though I expect some documentation explaining this to a reader. What should the user do? Replace this file with his/her own? Etc

miguelandres replied:
I'll add that when we are properly supporting this, right now I'm adding the backbone, but it's also useless to the user without the remote controller. 

When it's ready (and public) I will add these instructions to the readme (or start a Wiki since options start branching out).

pingpongboss commented on remixer_example/src/main/java/com/google/android/apps/remixer/RemixerApplication.java:
```diff
@@ -31,10 +32,21 @@
+      FirebaseRemoteControllerSyncer syncer =
+          new FirebaseRemoteControllerSyncer(getApplicationContext());
+      Remixer.getInstance().setSynchronizationMechanism(syncer);
+      syncer.startSyncing();
```
from API point of view, I would expect startSyncing() to be set up for me as long as I setSynchronizationMechanism() correctly.

miguelandres replied:
(I agree in principle, but there was a requirement of this having to be explicit, in a future PR this will be called from a UI control, also can be stopped from the UI.)

pingpongboss replied:
sgtm. Since the FirebaseRemoteControllerSyncer object is created here, what's the API like for the UI control to start and stop it? Will it have code like:

```
button.setOnClickListener( view -> {
  var syncer = Remixer.getInstance().getSynchronizationMechanism();
  ((FirebaseRemoteControllerSyncer) syncer).startSyncing();
});
```

pingpongboss commented on remixer_storage/proguard-rules.pro:
```diff
@@ -1,17 +1,9 @@
-# For more details, see
-#   http://developer.android.com/guide/developing/tools/proguard.html
+# Add this global rule
+-keepattributes Signature
```
what's this?

miguelandres replied:
Some firebase thing, it's on their (very unclear) documentation

pingpongboss replied:
that's fine. Just document it please. Link if possible.

miguelandres replied:
Done.

pingpongboss commented on remixer_storage/proguard-rules.pro:
```diff
@@ -1,17 +1,9 @@
-#}
+# This rule will properly ProGuard all the model classes in
+# the package com.yourcompany.models. Modify to fit the structure
+# of your app.
```
this documentation doesn't sound right

miguelandres replied:
Done

pingpongboss commented on remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/FirebaseRemoteControllerSyncer.java:
```diff
@@ -17,28 +17,105 @@
+  /**
+   * Initializes a {@code FirebaseRemoteControllerSyncer} instance.
+   *
+   * <p>Uses {@code context} to get
```
"to get"

miguelandres replied:
Hahaha done.

pingpongboss commented on remixer_example/src/main/java/com/google/android/apps/remixer/RemixerApplication.java:
```diff
@@ -31,10 +32,21 @@
-    Remixer.getInstance().setSynchronizationMechanism(new LocalStorage(getApplicationContext()));
+    if (USE_FIREBASE_REMOTE_CONTROLLER) {
+      FirebaseRemoteControllerSyncer syncer =
+          new FirebaseRemoteControllerSyncer(getApplicationContext());
```
is there a requirement that the context you pass in be the application context?

If so, I would recommend you pass in `this` here, and in new FirebaseRemoteControllerSyncer(), set `context = context.getApplicationContext()`.

That way, you're protected against any non-app contexts.

miguelandres replied:
Nah, no requirement. I could just do with this.

pingpongboss commented on remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/FirebaseRemoteControllerSyncer.java:
```diff
@@ -17,28 +17,105 @@
+   * <p>That way we get notified of changes to children (individual variables) after the initial
+   * sync.
+   */
+  public synchronized void startSyncing() {
```
also curiosu about whether `synchronized` is necessary here and below. Do you expect any of the synchronized methods to be called from different threads? Why?

miguelandres replied:
There is no guarantee of where the firebase callbacks will be called from, I do expect those to come from different threads.

pingpongboss commented on remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/FirebaseRemoteControllerSyncer.java:
```diff
@@ -49,27 +126,30 @@ public FirebaseRemoteControllerSyncer() {
    */
   private void clearVariablesInRemoteController() {
-    // TODO(miguely): write controller sync logic
+    reference.removeValue();
```
is this method still necessary? You can just inline `reference.removeValue()` (you actually call removeValue() twice in some cases)

miguelandres replied:
Done.

pingpongboss commented on remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/FirebaseRemoteControllerSyncer.java:
```diff
@@ -90,6 +173,46 @@ public void onContextRemoved(Object currentContext) {
+
+  // Implementation of ChildEventListener
+  @Override
+  public void onChildAdded(DataSnapshot dataSnapshot, String s) {
```
would like to see a better name than `s`

miguelandres replied:
Sorry, autogenerated code from firebase interface.

Commit:
Clean up after mark’s comments

- Rename arguments that had no meaningful name
- Complete Javadoc
- Remove one-liner method.
- improve proguard file
- Remove useless gitignore.

Commit:
Explain proguard rules.
